aiohttp_s3_client/credentials.py

```diff
@@ -0,0 +1,158 @@
+import configparser
+import json
+import os
+import socket
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Union
+from urllib.request import urlopen
+
+from yarl import URL
+
+
+@dataclass(frozen=True)
+class Credentials:
+    access_key_id: str
+    secret_access_key: str
+    region: str = ""
+    service: str = "s3"
+
+    def __bool__(self) -> bool:
+        return all((self.access_key_id, self.secret_access_key))
+
+    def __repr__(self) -> str:
+        return (
+            f"{self.__class__.__name__}(access_key_id={self.access_key_id!r}, "
+            "secret_access_key="
+            f'{"******" if self.secret_access_key else None!r}, '
+            f"region={self.region!r}, service={self.service!r})"
+        )
+
+    def as_dict(self) -> dict:
+        return {
+            "region": self.region,
+            "access_key_id": self.access_key_id,
+            "secret_access_key": self.secret_access_key,
+            "service": self.service,
+        }
+
+
+class URLCredentials(Credentials):
+    def __init__(
+        self, url: Union[str, URL], *, region: str = "", service: str = "s3"
+    ):
+        url = URL(url)
+        super().__init__(
+            access_key_id=url.user or '',
+            secret_access_key=url.password or '',
+            region=region, service=service,
+        )
+
+
+class EnvironmentCredentials(Credentials):
+    def __init__(self, region: str = '', service: str = "s3"):
+        super().__init__(
+            access_key_id=os.getenv("AWS_ACCESS_KEY_ID", ''),
+            secret_access_key=(
```

Comment:
same issue as below, need both secret_key and token

Reply:
I’m not an expert in aws, and not a customer btw, could you please provide the sanitized response examples, with header of course?
Should be fine for testing.

Reply:
$ curl -v 169.254.169.254/latest/meta-data/iam/security-credentials/s3devwriterole
*   Trying 169.254.169.254:80...
* Connected to 169.254.169.254 (169.254.169.254) port 80 (#0)
> GET /latest/meta-data/iam/security-credentials/myiamrole HTTP/1.1
> Host: 169.254.169.254
> User-Agent: curl/7.87.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: text/plain
< Accept-Ranges: none
< Last-Modified: Fri, 30 Jun 2023 19:06:40 GMT
< Content-Length: 1582
< Date: Fri, 30 Jun 2023 20:00:47 GMT
< Server: EC2ws
< Connection: close
< 
{
  "Code" : "Success",
  "LastUpdated" : "2023-06-30T19:06:42Z",
  "Type" : "AWS-HMAC",
  "AccessKeyId" : "ASIAV345345W4RX",
  "SecretAccessKey" : "VGJvJ5H345345345BQAJikpQN/Riq",
  "Token" : "IQoJb3JpZ2luX2VjEEQaC345345345kNBkUus5uk2a9JP3PXLhlo+0LKlEejP3cWkwMWARkyOCXPxoE7OD5XaH0em6trzY0ZCloPKH/yDCbkIKt9lecO/aYm+hSzrAFy",
  "Expiration" : "2023-07-01T01:25:35Z"
}
^^^^^ -- no trailing newline

GET /latest/meta-data/iam/security-credentials/ HTTP/1.1
> Host: 169.254.169.254
> User-Agent: curl/7.87.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: text/plain
< Accept-Ranges: none
< Last-Modified: Fri, 30 Jun 2023 19:06:40 GMT
< Content-Length: 14
< Date: Fri, 30 Jun 2023 20:02:18 GMT
< Server: EC2ws
< Connection: close
< 
myiamrole

^^^^^ -- no trailing newline



$ curl-v http://169.254.169.254/latest/dynamic/instance-identity/document
*   Trying 169.254.169.254:80...
* Connected to 169.254.169.254 (169.254.169.254) port 80 (#0)
> GET /latest/dynamic/instance-identity/document HTTP/1.1
> Host: 169.254.169.254
> User-Agent: curl/7.87.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: text/plain
< Accept-Ranges: none
< Last-Modified: Fri, 30 Jun 2023 19:06:40 GMT
< Content-Length: 478
< Date: Fri, 30 Jun 2023 20:03:14 GMT
< Server: EC2ws
< Connection: close
< 
{
  "accountId" : "123123",
  "architecture" : "x86_64",
  "availabilityZone" : "us-east-1a",
  "billingProducts" : null,
  "devpayProductCodes" : null,
  "marketplaceProductCodes" : null,
  "imageId" : "ami-123123",
  "instanceId" : "i-11232323",
  "instanceType" : "t3a.micro",
  "kernelId" : null,
  "pendingTime" : "2023-06-13T18:18:58Z",
  "privateIp" : "172.33.33.33",
  "ramdiskId" : null,
  "region" : "us-east-1",
  "version" : "2017-09-30"
}
^^^^^ -- no trailing newline